pennylane/data/attributes/operator/operator.py

```diff
@@ -0,0 +1,265 @@
+                # pennylane/ops/identity.py
+                qml.Identity,
+            )
+        )
+
+    def value_to_hdf5(self, bind_parent: HDF5Group, key: str, value: Op) -> HDF5Group:
+        return self._ops_to_hdf5(bind_parent, key, [value])
+
+    def hdf5_to_value(self, bind: HDF5Group) -> Op:
+        return self._hdf5_to_ops(bind)[0]
+
+    def _ops_to_hdf5(
+        self, bind_parent: HDF5Group, key: str, value: typing.Sequence[Operator]
+    ) -> HDF5Group:
+        """Serialize op sequence ``value``, and create nested sequences for any
+        composite ops in ``value``.
+
+        Since operators are commonly used in larger composite operations, we handle
+        sequences of operators as the default case. This allows for performant (in
+        time and space) serialization of large and nested operator sums, products, etc.
+        """
+        bind = bind_parent.create_group(key)
+
+        op_wire_labels = []
+        op_class_names = []
+        for i, op in enumerate(value):
+            op_key = f"op_{i}"
+            if type(op) not in self.consumes_types():
+                raise TypeError(
+                    f"Serialization of operator type {type(op).__name__} is not supported."
+                )
+
+            if isinstance(op, Tensor):
+                self._ops_to_hdf5(bind, op_key, op.obs)
+                op_wire_labels.append("null")
+            elif isinstance(op, qml.Hamiltonian):
+                coeffs, ops = op.terms()
+                ham_grp = self._ops_to_hdf5(bind, op_key, ops)
+                ham_grp["hamiltonian_coeffs"] = coeffs
+                op_wire_labels.append("null")
+            else:
+                bind[op_key] = op.data if len(op.data) else h5py.Empty("f")
+                op_wire_labels.append(wires_to_json(op.wires))
+
+            op_class_names.append(type(op).__name__)
+
+        bind["op_wire_labels"] = op_wire_labels
+        bind["op_class_names"] = op_class_names
+
+        return bind
+
+    def _hdf5_to_ops(self, bind: HDF5Group) -> List[Operator]:
+        """Load list of serialized ops from ``bind``."""
+        ops = []
+
+        op_class_names = bind["op_class_names"].asstr()
+        op_wire_labels = bind["op_wire_labels"].asstr()
+
+        for i, op_class_name in enumerate(op_class_names):
+            op_key = f"op_{i}"
+
+            op_cls = self._supported_ops_dict()[op_class_name]
+            if op_cls is Tensor:
+                ops.append(Tensor(*self._hdf5_to_ops(bind[op_key])))
+            elif op_cls is qml.Hamiltonian:
+                ops.append(
+                    qml.Hamiltonian(
+                        coeffs=list(bind[op_key]["hamiltonian_coeffs"]),
+                        observables=self._hdf5_to_ops(bind[op_key]),
+                    )
+                )
+            else:
+                wire_labels = json.loads(op_wire_labels[i])
+                op_data = bind[op_key]
+                if op_data.shape is not None:
```

Comment:
have you not hit any issues here? depending on the interface used (eg. none), `op.data` might not have a `shape` attribute:
```pycon
>>> qml.RX(1.1, 0).data.shape
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'shape'
```

`qml.math.shape` should work if it's needed though:
```pycon
>>> qml.math.shape(qml.RX(1.1, 0).data)
(1,)
```

tl;dr could you add tests that initialize a parametrized Operator like RX (eg. something other than Hamiltonian) using vanilla python values like 1.1 above, as well as using a numpy value that has a shape attribute?

Reply:
In this case, `op_data` is an HDF5 array, it always has a `shape` attribute. HDF5 arrays only have `shape=None` if they are explicitly defined as 'Empty'. This is distinct from a zero-length array:
```
>>> f = h5py.File("tmp", "w")
>>> f["empty"] = h5py.Empty("f")
>>> f["empty"].shape
None
>>> f["zero_length"] = []
>>> f["zero_length"].shape
(0,)
```

I'll add some tests for the RX operator